### PR TITLE
Improve payment service providers switching errors

### PR DESCRIPTION
**Pull request by luca-landa**
**Description**

This PR is a solution for https://github.com/solidusio/solidus/issues/3802.

Since old payment methods can't just be deleted after switching payment service provider, I provided a script to reset their `type` to a default `Spree::PaymentMethod`, to avoid breaking Single Table Inheritance while building `ActiveRecord` instances.
I added a `type_before_removal` field which will store the actual `type` value when the script is run. 

Also, the script sets `active`, `available_to_admin` and `available_to_users` flags to false.

Moreover, I replaced the typical  `ActiveRecord::SubclassNotFound` with a more meaningful error, which should save time to developers facing this issue and points them to the script.

If I got something wrong please tell me, this was my first experience with the project :)

**Checklist:**
- [x] I have followed [Pull Request guidelines](https://github.com/solidusio/solidus/blob/master/CONTRIBUTING.md#pull-request-guidelines)
- [x] I have added a detailed description into each commit message
- [x] I have updated Guides and README accordingly to this change (if needed)
- [x] I have added tests to cover this change (if needed)
- [x] I have attached screenshots to this PR for visual changes (if needed)

### CODE REVIEW

**Comment by luca-landa** [comment]
My bad, I didn't pay attention to failing checks.
I currently have no idea of what is causing the errors, I have run the full suite locally on postgresql without any trouble.
I will investigate and fix it.

**Comment by spaghetticode** [comment]
@luca-landa I'm wondering if we may be interested in keeping for reference the original payment method information.

For example, if we change the boolean field to a string, we can populate it with the previous `type` class name... what do you think?

**Comment by luca-landa** [comment]
@spaghetticode it seems a good idea to keep track of the previous `type`.

But personally I would add it in another field, instead of converting the `deprecated` field. 
I thought that the new flag could be useful to query for all current payment methods or for the deprecated ones. For these purposes, I wouldn't find it intuitive to write `.where.not(previous_type: nil)`, I would much rather do `.where(deprecated: false)` instead.

Do you think it can be a better solution?

**Comment by spaghetticode** [comment]
@luca-landa I don't have a strong opinion on this, I think there are pros and cons with both approaches. 

I don't see `where.not(previous_type: nil)` as a big issue, as it would probably just end up in a scope  such as
`scope :deprecated, -> {where.not(deprecated_type: nil)}`
Same goes for instance methods such as
```ruby
def deprecated?
  deprecated_type.present?
end
```
which would need to be explicitly written instead of relying on AR automatic definitions. The advantage in this case would be we just need to maintain a single field on the DB instead of 2.  Also, we already have the boolean field `active`, adding another one can potentially make things a bit messy when they're not kept in sync.

On the other hand, the burden to maintain 2 fields is rather minimal, and it would probably allow us to write simpler code in ruby.

**Comment by luca-landa** [comment]
@spaghetticode I agree on the potential problem of keeping `active` and `deprecated` in sync, I didn't really consider it at first and I'd like to avoid the risk.

I like your proposal with `deprecated_type` and the scope you defined (since `active` will help filter out deprecated records).

But I have another change in mind that would make the AR scope a bit more readable, and it would make the record more explicitly marked as "deprecated".
What if the new `type` set by the task is a new `Spree::PaymentMethod::DeprecatedMethod`model class, instead of `Spree::PaymentMethod`? The scope would also become `scope :deprecated, -> {where(type: "Spree::PaymentMethod::DeprecatedMethod")}`.

I have another (OT) question about this PR: I don't know how exactly should I fix an error that I made, I could add another commit but it wouldn't satisfy the guidelines request that the specs must pass for each commit. Should I create another PR for another branch?

File: core/db/migrate/20201108142424_add_deprecated_to_payment_method.rb
Comment (jarednorman) [inline core/db/migrate/20201108142424_add_deprecated_to_payment_method.rb:5]: I don't think "deprecated" is the right name for this field. "Deprecated" means "usable but regarded as obsolete and best avoided, typically due to having been superseded". I think in this case the payment method will no longer be usable, so we should find a better way to describe it.

**Review by jarednorman — CHANGES_REQUESTED** [review]
I think we should find a better name for the "deprecated" column that aligns better with how it works. Otherwise, this seems like a reasonable approach if the method we're overriding has the desired effect.

File: core/lib/tasks/payment_method.rake
Comment (kennyadsl) [inline core/lib/tasks/payment_method.rake:19]: Why use `.name` instead of a string? Not saying it's wrong, I just would like to understand the reasoning behind it.

File: core/lib/tasks/payment_method.rake
Comment (kennyadsl) [inline core/lib/tasks/payment_method.rake:28]: Is there a reason why we can't set the type here as well?

File: core/lib/tasks/payment_method.rake
Comment (kennyadsl) [inline core/lib/tasks/payment_method.rake:7]: Maybe I'm missing something but what about removing these methods and just leave the code to execute the change when the exception is raised? Something like:

```ruby
    Spree::PaymentMethod.pluck(:type).select do |type|
      begin
        ActiveSupport::Dependencies.constantize(type)
      rescue NameError
        # do the updates here
      end
    end
```

File: core/app/models/spree/payment_method.rb
Comment (kennyadsl) [inline core/app/models/spree/payment_method.rb:14]: I think this could be a good spot for https://api.rubyonrails.org/classes/ActiveSupport/ActionableError/ClassMethods.html, which is the same class used to run migrations directly in the browser when needed. What do you think?

File: guides/source/developers/payments/payment-service-providers.html.md
Comment (kennyadsl) [inline guides/source/developers/payments/payment-service-providers.html.md:68]: 👍 adding this documentation, thanks!

**Review by kennyadsl — COMMENTED** [review]
Thanks! I left some questions on the implementation. Looking forward to know what you think about them. 

 Also, can you please squash the last commit into the one that introduced the code style issue?

🙏

File: core/db/migrate/20201108142424_add_deprecated_to_payment_method.rb
Comment (luca-landa) [reply core/db/migrate/20201108142424_add_deprecated_to_payment_method.rb:5]: As I said discussing with @spaghetticode, I'd like the db row to be clearly marked as "deprecated" (or another term). I didn't understand from the solidus docs in which cases should a payment method be flagged with `active = false`, and nor if there are cases where it makes sense to set it to `false` at some point and again to `true` after a while. So, I wasn't really sure that the `active` could be used for this case.

Given that, if you think that `active` can be used to express this state I would be glad to remove the new flag, or even better to change it into `previous_type` as suggested from @spaghetticode to keep track of the previous real type of the payment method.

File: core/lib/tasks/payment_method.rake
Comment (luca-landa) [reply core/lib/tasks/payment_method.rake:19]: I usually prefer to reference constants or variables when possible, just in case they may change in the future (even tough `Spree::PaymentMethod` is very unlikely to be renamed). Even having the tests, being able to reference a class to fix an error that requires that field to contain a valid class name seemed a bit safer to me.

File: core/lib/tasks/payment_method.rake
Comment (luca-landa) [reply core/lib/tasks/payment_method.rake:28]: That's the point of this fix actually: even running a query `Spree::PaymentMethod.where(...)` before changing the `type` via raw sql would make AR immediately raise the STI error, while creating the result instances, so `type` needs to be changed before that.
So, I could have updated the record fully in the first raw sql update, but I preferred to separate this second update to let AR handle boolean values via its connection adapter.

File: core/lib/tasks/payment_method.rake
Comment (luca-landa) [reply core/lib/tasks/payment_method.rake:7]: I'm a bit confused by what you quoted, I'll try to answer for what I understood.
I personally wouldn't apply the fix automatically when the STI error happens: that error reports an inconsistent situation on the db and fixing it actually changes a bit of the behavior of the application (querying for payment methods will return instances of another class after that).
I don't like the idea of this happening without the store devs understanding this problem, and I thought that they may need to take more actions than just this when this problem occurs.

But I lack of experience on this project, so please correct me if my assumption is wrong.

About the code you inserted: that rescue instruction is where the task understands what types are not supported anymore, and not where the original error is caught to raise a more understandable error.
The error is caught in `Spree::PaymentMethod#find_sti_class,` so it should be caught there in order to make this automatic. Did I understand correctly what you meant?

File: core/app/models/spree/payment_method.rb
Comment (luca-landa) [reply core/app/models/spree/payment_method.rb:14]: It is a good idea, I actually didn't think about it at first (and didn't know about it until now). 

But that's a Rails 6 feature and Solidus minimum requirement is for Rails 5.0.0, so it can't be implemented, am I right?

**Comment by spaghetticode** [comment]
> But I have another change in mind that would make the AR scope a bit more readable, and it would make the record more explicitly marked as "deprecated".
> What if the new `type` set by the task is a new `Spree::PaymentMethod::DeprecatedMethod`model class, instead of `Spree::PaymentMethod`? The scope would also become `scope :deprecated, -> {where(type: "Spree::PaymentMethod::DeprecatedMethod")}`.

I guess that would make sense, although it's not strictly needed... but I think it may make it easier to discriminate _removed_ payment methods and their payments? If while doing the PR you can come up with good reasons for doing (or not doing) it, you can add some comments in the commit message, that would be very valuable.

> I have another (OT) question about this PR: I don't know how exactly should I fix an error that I made, I could add another commit but it wouldn't satisfy the guidelines request that the specs pass for each commit. Should I create another PR for another branch?

No need, you can make all the fixes and changes you need in this PR, as long as they're related to it. 
You can also rebase, do whatever makes sense so that, eventually, commits make sense when we'll need to navigate the git history and understand the changes.

**Comment by luca-landa** [comment]
Well the `Spree::PaymentMethod::DeprecatedMethod` class was just a proposal, I don't see any more good reasons than those I already explained, but I actually see a reason for not doing it: we would be adding a subclass which does not define any specific behavior.

I may just discard the idea, but stick with the idea of changing `deprecated` to `previous_type` if the other reviewers approve this change too.

File: core/app/models/spree/payment_method.rb
Comment (kennyadsl) [reply core/app/models/spree/payment_method.rb:14]: Yes, you are right, but if we like the idea we could conditionally include `ActiveSupport::ActionableError` and its related code only when that class is defined. Up to you!

File: core/db/migrate/20201108142424_add_deprecated_to_payment_method.rb
Comment (kennyadsl) [reply core/db/migrate/20201108142424_add_deprecated_to_payment_method.rb:5]: These concerns make sense and I don't have a clear answer as well. Just thinking aloud: is there any occurrence when we want `active` and `deprecated` to have different values at the same time. I mean does it make sense having `deprecated: true` without also having `active: false`? I don't think so, since the payment method class has been removed from the project, having that payment method active would just produce this error more often. This makes me think that active is just fine for this purpose as well. 

If we also add `previous_type`, this new field populated would be enough to understand which records are in this specific state.

But again, I don't have strong opinions here and I'd leave this decision to you. I'm fine with both as soon as we don't use the `deprecated` term for the reasons above.

File: core/lib/tasks/payment_method.rake
Comment (kennyadsl) [reply core/lib/tasks/payment_method.rake:7]: Sorry, my comment wasn't clear at all: I meant to do something like this directly in the rake task:

```ruby
  task deprecate_unsupported_payment_methods: :environment do
    Spree::PaymentMethod.pluck(:id, :type).each do |id, type|
      begin
        ActiveSupport::Dependencies.constantize(type)
      rescue NameError
        # we have the id of the problematic type here, can we avoid the extra query and just update the record?
      end
    end
  end
```

File: core/lib/tasks/payment_method.rake
Comment (kennyadsl) [reply core/lib/tasks/payment_method.rake:28]: Sure, you are right. I'd try to make a single query though, maybe we can move these attributes settings to the raw SQL update?

**Review by kennyadsl — COMMENTED** [review]
Left a second review with other thoughts. Let me know if those make sense. 

Another thing: I was wondering if there's something we could do when deleting a payment method. I'm thinking of a simple PORO that handles the PaymentMethod destruction and maybe sets these attributes for us when we (soft-)delete a payment method? This way we will also prevent this error from happening, anyway, maybe it's something for another PR. 🙂

File: core/db/migrate/20201108142424_add_deprecated_to_payment_method.rb
Comment (luca-landa) [reply core/db/migrate/20201108142424_add_deprecated_to_payment_method.rb:5]: Then adding `previous_type` and removing `deprecated` seems a good choice. I'm just not 100% sure on the name, can you think of something more expressive and not too long? Maybe `removed_type`?

File: core/lib/tasks/payment_method.rake
Comment (luca-landa) [reply core/lib/tasks/payment_method.rake:28]: Updating in a single query (and every single row individually, as you suggested below), would make the code something like this.
```ruby
def update_row(id)
    connection = ActiveRecord::Base.connection
    false_value = connection.adapter_name[/sqlite/i] ? '0' : 'false'
    true_value = connection.adapter_name[/sqlite/i] ? '1' : 'true'
    connection.exec_query(
      "UPDATE spree_payment_methods "\
      "SET type='#{Spree::PaymentMethod.name}', deprecated=#{true_value}, "\
      "active=#{false_value}, available_to_users=#{false_value}, available_to_admin=#{false_value} "\
      "WHERE id=#{id}"
    )
end
```
The task becomes really shorter, I like it, but I don't find really nice to manually choose what to write depending on the db in use (it's a db adapter's job), since doing it via AR would be more stable (for example, in case another db will be supported in the future that another "true" value).
Do you think I'm just worrying too much? It's unlikely to support another db that will require another boolean value different from "0"/"1" and "false"/"true", isn't it?

File: core/lib/tasks/payment_method.rake
Comment (luca-landa) [reply core/lib/tasks/payment_method.rake:7]: It's possible, the records will be a few, so I can see no performance issue. It would make much more sense merging the two update statements in the first one, as you asked above. I'll wait for you to answer to that before changing this.

File: core/app/models/spree/payment_method.rb
Comment (luca-landa) [reply core/app/models/spree/payment_method.rb:14]: Just to be sure, you mean like a module to include that could include all the stuff related to this PR? Including a static method to do the actual updates, the `ActiveSupport::ActionableError` and maybe the overriding of `find_sti_class` to catch the actual error?

File: core/lib/tasks/payment_method.rake
Comment (spaghetticode) [reply core/lib/tasks/payment_method.rake:28]: ActiveRecord provides facilities for avoiding hardcoding those values, you can use `ActiveRecord::Base.connection.quoted_true` and `ActiveRecord::Base.connection.quoted_false`, this should make the SQL generic enough IMO.

File: core/app/models/spree/payment_method.rb
Comment (kennyadsl) [reply core/app/models/spree/payment_method.rb:14]: I mean something like:

```ruby
class UnsupportedPaymentMethod < StandardError
  if defined?(ActiveSupport::ActionableError)
    include ActiveSupport::ActionableError

    action "Bla bla" do
      Our::Rake::Task.execute
    end
  end
end

File: core/db/migrate/20201108142424_add_deprecated_to_payment_method.rb
Comment (luca-landa) [reply core/db/migrate/20201108142424_add_deprecated_to_payment_method.rb]: I changed `deprecated` to `previous_type` and updated the doc accordingly.

File: core/lib/tasks/payment_method.rake
Comment (luca-landa) [reply core/lib/tasks/payment_method.rake]: Thanks! I couldn't find it myself, this helped me changing the script as suggested by @kennyadsl, I just force pushed the change.

File: core/app/models/spree/payment_method.rb
Comment (luca-landa) [reply core/app/models/spree/payment_method.rb:14]: I tried to implement it without success: the button doesn't appear if the error was raised in `ActiveRecord.find_sti_class`. I could only find base examples of this feature usage and nothing in official docs that helped my solve my specific case. I would leave this for another PR for now (and maybe I'd open an issue after a bit more digging).

File: core/db/migrate/20201118200630_add_previous_type_to_spree_payment_methods.rb
Comment (kennyadsl) [inline core/db/migrate/20201118200630_add_previous_type_to_spree_payment_methods.rb:5]: What about `type_before_removal`? I think it expresses better what this field contains.

File: core/lib/tasks/payment_method.rake
Comment (kennyadsl) [inline core/lib/tasks/payment_method.rake:22]: Not mandatory, but I'd move all these attributes to separate lines. This will allow us to add/change/remove individual lines in the future without impacting git history for all of them.

File: guides/source/developers/payments/payment-service-providers.html.md
Comment (kennyadsl) [inline guides/source/developers/payments/payment-service-providers.html.md:72]: I'd specify that this prevents the system to raise the `ActiveRecord::SubclassNotFound` error. This way, it will also be easier to find if someone is searching using the exception name as a keyword.

**Review by kennyadsl — COMMENTED** [review]
Left some other comments, we are almost there, thanks again!

File: core/db/migrate/20201118200630_add_previous_type_to_spree_payment_methods.rb
Comment (luca-landa) [reply core/db/migrate/20201118200630_add_previous_type_to_spree_payment_methods.rb]: It sounds better actually, fixed.

File: core/lib/tasks/payment_method.rake
Comment (luca-landa) [reply core/lib/tasks/payment_method.rake]: Fixed it, actually it was a bit ugly before.

File: guides/source/developers/payments/payment-service-providers.html.md
Comment (luca-landa) [reply guides/source/developers/payments/payment-service-providers.html.md:74]: Nice suggestion, I didn't consider it at first. I just added it.

File: core/spec/models/spree/payment_method_spec.rb
Comment (spaghetticode) [inline core/spec/models/spree/payment_method_spec.rb:404]: I think this `before` block has no effect on the record, as it's updating the type but the value is the same as the initial one from line 402.

Comment (luca-landa) [reply core/spec/models/spree/payment_method_spec.rb]: Right, I removed the block.

**Comment by spaghetticode** [comment]
@luca-landa the PR is almost ready now, thanks for the work you've done so far! There are still a couple of cosmetic issues raised by Hound, and this [merge commit](https://github.com/luca-landa/solidus/commit/fb03992a0fbd0212d9f1935ba414700acb792007) that should be dealt with.

**Comment by luca-landa** [comment]
@spaghetticode I'm sorry I completely lost track of this PR 😅 it should be fixed now

**Review by spaghetticode — APPROVED** [review]
LGTM, thanks Luca!

**Comment by waiting-for-dev** [comment]
Hey @jarednorman, I think [your concern](https://github.com/solidusio/solidus/pull/3837#pullrequestreview-530851981) was addressed :slightly_smiling_face:  Is there anything else we need to tackle here?

**Review by jarednorman — APPROVED** [review]
No, I'm happy here.

**Pull request merged**
